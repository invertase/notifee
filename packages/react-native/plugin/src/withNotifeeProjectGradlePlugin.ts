import { ConfigPlugin, WarningAggregator, withProjectBuildGradle } from '@expo/config-plugins';

const withNotifeeProjectGradlePlugin: ConfigPlugin = config => {
  return withProjectBuildGradle(config, ({ modResults, ...subConfig }) => {
    if (modResults.language !== 'groovy') {
      WarningAggregator.addWarningAndroid(
        'withNotifee',
        `Cannot automatically configure project build.gradle if it's not groovy`,
      );
      return { modResults, ...subConfig };
    }

    modResults.contents = setMavenRepository(modResults.contents);
    return { modResults, ...subConfig };
  });
};

const setMavenRepository = (projectBuildGradle: string): string => {
  if (projectBuildGradle.includes('@notifee/react-native/android/libs')) return projectBuildGradle;

  return projectBuildGradle.replace(
    /mavenLocal\(\)/,
    `mavenLocal()
        maven { url "$rootDir/../node_modules/@notifee/react-native/android/libs" }`,
  );
};

export { setMavenRepository };
export default withNotifeeProjectGradlePlugin;
