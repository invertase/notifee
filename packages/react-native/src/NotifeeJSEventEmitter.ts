/*
 * Copyright (c) 2016-present Invertase Limited
 */

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import EventEmitter from 'react-native/Libraries/vendor/emitter/EventEmitter';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore See https://github.com/facebook/react-native/pull/36462
const emitter = new EventEmitter();
export default emitter;
