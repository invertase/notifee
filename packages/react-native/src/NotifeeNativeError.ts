/*
 * Copyright (c) 2016-present Invertase Limited
 */

import { NativeError } from './types/Library';

export default class NotifeeNativeError extends Error implements NativeError {
  public readonly code: string;
  public readonly nativeErrorCode: string;
  public readonly nativeErrorMessage: string;
  private readonly jsStack: string;

  //  TODO native error type
  constructor(nativeError: any, jsStack = '') {
    super();
    const { userInfo } = nativeError;

    this.code = `${userInfo.code || 'unknown'}`;
    Object.defineProperty(this, 'code', {
      enumerable: false,
      value: `${userInfo.code || 'unknown'}`,
    });

    this.jsStack = jsStack;
    Object.defineProperty(this, 'jsStack', {
      enumerable: false,
      value: jsStack,
    });

    Object.defineProperty(this, 'message', {
      enumerable: false,
      value: `[${this.code}] ${userInfo.message || nativeError.message}`,
    });

    Object.defineProperty(this, 'userInfo', {
      enumerable: false,
      value: userInfo,
    });

    this.nativeErrorCode = userInfo.nativeErrorCode || null;
    Object.defineProperty(this, 'nativeErrorCode', {
      enumerable: false,
      value: userInfo.nativeErrorCode || null,
    });

    this.nativeErrorMessage = userInfo.nativeErrorMessage || null;
    Object.defineProperty(this, 'nativeErrorMessage', {
      enumerable: false,
      value: userInfo.nativeErrorMessage || null,
    });

    this.stack = this.getStackWithMessage(`NativeError: ${this.message}`);
  }

  // todo errorEvent type
  static fromEvent(errorEvent: any, stack?: string): NotifeeNativeError {
    return new NotifeeNativeError({ userInfo: errorEvent }, stack || new Error().stack);
  }

  /**
   * Build a stack trace that includes JS stack prior to calling the native method.
   *
   * @returns {string}
   */
  getStackWithMessage(message: string): string {
    return [message, ...this.jsStack.split('\n').slice(2, 13)].join('\n');
  }
}
