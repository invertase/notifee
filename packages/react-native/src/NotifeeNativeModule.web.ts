import { EventEmitter, NativeEventEmitter, NativeModulesStatic } from 'react-native';
import { AuthorizationStatus, Notification } from './types/Notification';

export interface NativeModuleConfig {
  version: string;
  nativeModuleName: string;
  nativeEvents: string[];
}

export default class NotifeeNativeModule {
  private readonly _moduleConfig: NativeModuleConfig;

  private sw: ServiceWorkerRegistration | undefined;

  public constructor(config: NativeModuleConfig) {
    this._moduleConfig = Object.assign({}, config);

    if (NotifeeNativeModule.hasServiceWorkerSupport) {
      navigator.serviceWorker
        .register('notifee-sw.js')
        .then(serviceWorkerRegistration => {
          this.sw = serviceWorkerRegistration;
          console.log('Service worker "notifee-sw.js" is registered.');
        })
        .catch(e => {
          console.error(e);
          console.log(
            'The service worker could not be registered. Using the browser Notification.\n' +
            'Browser Notification does not include all features, to unlock Notifee full power add `notifee-sw.js` service worker.\n' +
            'Learn how to add "notifee-sw.js": https://notifee.com',
          );
        });
    } else {
      console.log('Your browser does not support service workers. Using the browser Notification.');
    }
  }

  public get emitter(): EventEmitter {
    return new NativeEventEmitter();
  }

  public get native(): NativeModulesStatic {
    const sw = this.sw;
    const hasNotificationSupport = NotifeeNativeModule.hasNotificationSupport;
    const formatNotificationBody = NotifeeNativeModule.formatNotificationBody;

    return {
      requestPermission: (): Promise<AuthorizationStatus> => {
        if (!hasNotificationSupport) return Promise.resolve(AuthorizationStatus.NOT_DETERMINED);

        return window.Notification.requestPermission().then(permission => {
          switch (permission) {
            case 'default':
              return AuthorizationStatus.NOT_DETERMINED;
            case 'denied':
              return AuthorizationStatus.DENIED;
            case 'granted':
              return AuthorizationStatus.AUTHORIZED;
          }
        });
      },
      displayNotification: (notification: Notification): Promise<void> => {
        if (sw) {
          return sw.showNotification(notification.title ?? '', {
            body: formatNotificationBody(notification.subtitle, notification.body),
            data: notification.data,
          });
        } else if (hasNotificationSupport) {
          new Notification(notification.title ?? '', {
            body: formatNotificationBody(notification.subtitle, notification.body),
            data: notification.data,
          });
        }
        return Promise.resolve();
      },
    };
  }

  private static get hasServiceWorkerSupport() {
    return 'serviceWorker' in navigator;
  }

  private static get hasNotificationSupport() {
    return 'Notification' in window;
  }

  private static formatNotificationBody(subtitle?: string, body?: string) {
    let txt = '';
    if (subtitle) txt += subtitle;
    if (subtitle && body) txt += '\n';
    if (body) txt += body;
    return txt;
  }
}
