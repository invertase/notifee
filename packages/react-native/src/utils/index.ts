/*
 * Copyright (c) 2016-present Invertase Limited
 */

import { Platform } from 'react-native';

export * from './id';
export * from './validate';

export function isError(value: object): boolean {
  if (Object.prototype.toString.call(value) === '[object Error]') {
    return true;
  }

  return value instanceof Error;
}

export function objectHasProperty<T>(
  target: T,
  property: string | number | symbol,
): property is keyof T {
  return Object.hasOwnProperty.call(target, property);
}

export const isIOS = Platform.OS === 'ios';

export const isAndroid = Platform.OS === 'android';

export const isWeb = Platform.OS === 'web';

export function noop(): void {
  // noop-üêà
}
export const kReactNativeNotifeeForegroundServiceHeadlessTask =
  'app.notifee.foreground-service-headless-task';

export const kReactNativeNotifeeNotificationEvent = 'app.notifee.notification-event';

export const kReactNativeNotifeeNotificationBackgroundEvent =
  'app.notifee.notification-event-background';

export enum NotificationType {
  ALL = 0,
  DISPLAYED = 1,
  TRIGGER = 2,
}
