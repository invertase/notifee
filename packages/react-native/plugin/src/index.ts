import { ConfigPlugin, withPlugins } from '@expo/config-plugins';

import withNotifeeProjectGradlePlugin from './withNotifeeProjectGradlePlugin';

const withNotifee: ConfigPlugin = config => {
  return withPlugins(config, [withNotifeeProjectGradlePlugin]);
};

export default withNotifee;
