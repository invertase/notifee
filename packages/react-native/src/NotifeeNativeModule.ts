/*
 * Copyright (c) 2016-present Invertase Limited
 */

import NotifeeJSEventEmitter from './NotifeeJSEventEmitter';
import {
  EventSubscriptionVendor,
  NativeEventEmitter,
  NativeModules,
  NativeModulesStatic,
} from 'react-native';

export interface NativeModuleConfig {
  version: string;
  nativeModuleName: string;
  nativeEvents: string[];
}

export default class NotifeeNativeModule {
  private readonly _moduleConfig: NativeModuleConfig;
  private _nativeModule: NativeModulesStatic | null;
  private _nativeEmitter: NativeEventEmitter;

  public constructor(config: NativeModuleConfig) {
    this._nativeModule = null;
    this._moduleConfig = Object.assign({}, config);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore - change here needs resolution https://github.com/DefinitelyTyped/DefinitelyTyped/pull/49560/files
    this._nativeEmitter = new NativeEventEmitter(this.native as EventSubscriptionVendor);
    for (let i = 0; i < config.nativeEvents.length; i++) {
      const eventName = config.nativeEvents[i];
      this._nativeEmitter.addListener(eventName, (payload: any) => {
        this.emitter.emit(eventName, payload);
      });
    }
  }

  public get emitter() {
    return NotifeeJSEventEmitter;
  }

  public get native(): NativeModulesStatic {
    if (this._nativeModule) {
      return this._nativeModule;
    }

    this._nativeModule = NativeModules[this._moduleConfig.nativeModuleName];
    if (this._nativeModule == null) {
      throw new Error('Notifee native module not found.');
    }

    return this._nativeModule;
  }
}
