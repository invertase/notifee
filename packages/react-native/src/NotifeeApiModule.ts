/*
 * Copyright (c) 2016-present Invertase Limited
 */

import { AppRegistry, Platform } from 'react-native';
import { Module } from './types/Module';
import {
  AndroidChannel,
  AndroidChannelGroup,
  AndroidNotificationSetting,
  NativeAndroidChannel,
  NativeAndroidChannelGroup,
} from './types/NotificationAndroid';
import {
  AuthorizationStatus,
  InitialNotification,
  Notification,
  Event,
  TriggerNotification,
  DisplayedNotification,
  NotificationSettings,
} from './types/Notification';
import { PowerManagerInfo } from './types/PowerManagerInfo';
import { Trigger } from './types/Trigger';
import NotifeeNativeModule, { NativeModuleConfig } from './NotifeeNativeModule';
import {
  isAndroid,
  isArray,
  isFunction,
  isIOS,
  isNumber,
  isString,
  isUndefined,
  isWeb,
  kReactNativeNotifeeForegroundServiceHeadlessTask,
  kReactNativeNotifeeNotificationBackgroundEvent,
  kReactNativeNotifeeNotificationEvent,
  NotificationType,
} from './utils';
import validateNotification from './validators/validateNotification';
import validateTrigger from './validators/validateTrigger';
import validateAndroidChannel from './validators/validateAndroidChannel';
import validateAndroidChannelGroup from './validators/validateAndroidChannelGroup';
import { IOSNotificationCategory, IOSNotificationPermissions } from './types/NotificationIOS';
import validateIOSCategory from './validators/validateIOSCategory';
import validateIOSPermissions from './validators/validateIOSPermissions';

let backgroundEventHandler: (event: Event) => Promise<void>;

let registeredForegroundServiceTask: (notification: Notification) => Promise<void>;

if (isAndroid) {
  // Register foreground service
  AppRegistry.registerHeadlessTask(kReactNativeNotifeeForegroundServiceHeadlessTask, () => {
    if (!registeredForegroundServiceTask) {
      console.warn(
        '[notifee] no registered foreground service has been set for displaying a foreground notification.',
      );
      return (): Promise<void> => Promise.resolve();
    }

    return ({ notification }): Promise<void> => registeredForegroundServiceTask(notification);
  });
}

export default class NotifeeApiModule extends NotifeeNativeModule implements Module {
  constructor(config: NativeModuleConfig) {
    super(config);
    if (isAndroid) {
      // Register background handler
      AppRegistry.registerHeadlessTask(kReactNativeNotifeeNotificationEvent, () => {
        return (event: Event): Promise<void> => {
          if (!backgroundEventHandler) {
            console.warn(
              '[notifee] no background event handler has been set. Set a handler via the "onBackgroundEvent" method.',
            );
            return Promise.resolve();
          }
          return backgroundEventHandler(event);
        };
      });
    } else if (isIOS) {
      this.emitter.addListener(
        kReactNativeNotifeeNotificationBackgroundEvent,
        (event: Event): Promise<void> => {
          if (!backgroundEventHandler) {
            console.warn(
              '[notifee] no background event handler has been set. Set a handler via the "onBackgroundEvent" method.',
            );
            return Promise.resolve();
          }

          return backgroundEventHandler(event);
        },
      );
    }
  }

  public getTriggerNotificationIds = (): Promise<string[]> => {
    if (isAndroid || isIOS) {
      return this.native.getTriggerNotificationIds();
    }

    return Promise.resolve([]);
  };

  public getTriggerNotifications = (): Promise<TriggerNotification[]> => {
    if (isAndroid || isIOS) {
      return this.native.getTriggerNotifications();
    }

    return Promise.resolve([]);
  };

  public getDisplayedNotifications = (): Promise<DisplayedNotification[]> => {
    if (isAndroid || isIOS) {
      return this.native.getDisplayedNotifications();
    }

    return Promise.resolve([]);
  };

  public isChannelBlocked = (channelId: string): Promise<boolean> => {
    if (!isString(channelId)) {
      throw new Error("notifee.isChannelBlocked(*) 'channelId' expected a string value.");
    }

    if (isWeb || isIOS || this.native.ANDROID_API_LEVEL < 26) {
      return Promise.resolve(false);
    }

    return this.native.isChannelBlocked(channelId);
  };

  public isChannelCreated = (channelId: string): Promise<boolean> => {
    if (!isString(channelId)) {
      channelId;
      throw new Error("notifee.isChannelCreated(*) 'channelId' expected a string value.");
    }

    if (isWeb || isIOS || this.native.ANDROID_API_LEVEL < 26) {
      return Promise.resolve(true);
    }

    return this.native.isChannelCreated(channelId);
  };

  public cancelAllNotifications = (notificationIds?: string[], tag?: string): Promise<void> => {
    if (isAndroid || isIOS) {
      if (notificationIds) {
        if (isAndroid) {
          return this.native.cancelAllNotificationsWithIds(
            notificationIds,
            NotificationType.ALL,
            tag,
          );
        }
        return this.native.cancelAllNotificationsWithIds(notificationIds);
      }
      return this.native.cancelAllNotifications();
    }

    return Promise.resolve();
  };

  public cancelDisplayedNotifications = (
    notificationIds?: string[],
    tag?: string,
  ): Promise<void> => {
    if (isAndroid || isIOS) {
      if (notificationIds) {
        if (isAndroid) {
          return this.native.cancelAllNotificationsWithIds(
            notificationIds,
            NotificationType.DISPLAYED,
            tag,
          );
        }

        return this.native.cancelDisplayedNotificationsWithIds(notificationIds);
      }

      return this.native.cancelDisplayedNotifications();
    }

    return Promise.resolve();
  };

  public cancelTriggerNotifications = (notificationIds?: string[]): Promise<void> => {
    if (isAndroid || isIOS) {
      if (notificationIds) {
        if (isAndroid) {
          return this.native.cancelAllNotificationsWithIds(
            notificationIds,
            NotificationType.TRIGGER,
            null,
          );
        }
        return this.native.cancelTriggerNotificationsWithIds(notificationIds);
      }
      return this.native.cancelTriggerNotifications();
    }

    return Promise.resolve();
  };

  public cancelNotification = (notificationId: string, tag?: string): Promise<void> => {
    if (!isString(notificationId)) {
      throw new Error("notifee.cancelNotification(*) 'notificationId' expected a string value.");
    }

    if (isAndroid) {
      return this.native.cancelAllNotificationsWithIds([notificationId], NotificationType.ALL, tag);
    }

    if (isIOS) {
      return this.native.cancelNotification(notificationId);
    }

    return Promise.resolve();
  };

  public cancelDisplayedNotification = (notificationId: string, tag?: string): Promise<void> => {
    if (!isString(notificationId)) {
      throw new Error(
        "notifee.cancelDisplayedNotification(*) 'notificationId' expected a string value.",
      );
    }

    if (isAndroid) {
      return this.native.cancelAllNotificationsWithIds(
        [notificationId],
        NotificationType.DISPLAYED,
        tag,
      );
    }

    if (isIOS) {
      return this.native.cancelDisplayedNotification(notificationId);
    }

    return Promise.resolve();
  };

  public cancelTriggerNotification = (notificationId: string): Promise<void> => {
    if (!isString(notificationId)) {
      throw new Error(
        "notifee.cancelTriggerNotification(*) 'notificationId' expected a string value.",
      );
    }

    if (isAndroid) {
      return this.native.cancelAllNotificationsWithIds(
        [notificationId],
        NotificationType.TRIGGER,
        null,
      );
    }

    if (isIOS) {
      return this.native.cancelTriggerNotification(notificationId);
    }

    return Promise.resolve();
  };

  public createChannel = (channel: AndroidChannel): Promise<string> => {
    let options: AndroidChannel;
    try {
      options = validateAndroidChannel(channel);
    } catch (e: any) {
      throw new Error(`notifee.createChannel(*) ${e.message}`);
    }

    if (isAndroid) {
      if (this.native.ANDROID_API_LEVEL < 26) {
        return Promise.resolve(options.id);
      }

      return this.native.createChannel(options).then(() => {
        return options.id;
      });
    }

    return Promise.resolve('');
  };

  public createChannels = (channels: AndroidChannel[]): Promise<void> => {
    if (!isArray(channels)) {
      throw new Error("notifee.createChannels(*) 'channels' expected an array of AndroidChannel.");
    }

    const options: AndroidChannel[] = [];
    try {
      for (let i = 0; i < channels.length; i++) {
        options[i] = validateAndroidChannel(channels[i]);
      }
    } catch (e: any) {
      throw new Error(`notifee.createChannels(*) 'channels' a channel is invalid: ${e.message}`);
    }

    if (isAndroid && this.native.ANDROID_API_LEVEL >= 26) {
      return this.native.createChannels(options);
    }

    return Promise.resolve();
  };

  public createChannelGroup = (channelGroup: AndroidChannelGroup): Promise<string> => {
    let options: AndroidChannelGroup;
    try {
      options = validateAndroidChannelGroup(channelGroup);
    } catch (e: any) {
      throw new Error(`notifee.createChannelGroup(*) ${e.message}`);
    }

    if (isAndroid) {
      if (this.native.ANDROID_API_LEVEL < 26) {
        return Promise.resolve(options.id);
      }

      return this.native.createChannelGroup(options).then(() => {
        return options.id;
      });
    }

    return Promise.resolve('');
  };

  public createChannelGroups = (channelGroups: AndroidChannelGroup[]): Promise<void> => {
    if (!isArray(channelGroups)) {
      throw new Error(
        "notifee.createChannelGroups(*) 'channelGroups' expected an array of AndroidChannelGroup.",
      );
    }

    const options = [];
    try {
      for (let i = 0; i < channelGroups.length; i++) {
        options[i] = validateAndroidChannelGroup(channelGroups[i]);
      }
    } catch (e: any) {
      throw new Error(
        `notifee.createChannelGroups(*) 'channelGroups' a channel group is invalid: ${e.message}`,
      );
    }

    if (isAndroid && this.native.ANDROID_API_LEVEL >= 26) {
      return this.native.createChannelGroups(options);
    }

    return Promise.resolve();
  };

  public deleteChannel = (channelId: string): Promise<void> => {
    if (!isString(channelId)) {
      throw new Error("notifee.deleteChannel(*) 'channelId' expected a string value.");
    }

    if (isAndroid && this.native.ANDROID_API_LEVEL >= 26) {
      return this.native.deleteChannel(channelId);
    }

    return Promise.resolve();
  };

  public deleteChannelGroup = (channelGroupId: string): Promise<void> => {
    if (!isString(channelGroupId)) {
      throw new Error("notifee.deleteChannelGroup(*) 'channelGroupId' expected a string value.");
    }

    if (isAndroid && this.native.ANDROID_API_LEVEL >= 26) {
      return this.native.deleteChannelGroup(channelGroupId);
    }

    return Promise.resolve();
  };

  public displayNotification = (notification: Notification): Promise<string> => {
    let options: Notification;
    try {
      options = validateNotification(notification);
    } catch (e: any) {
      throw new Error(`notifee.displayNotification(*) ${e.message}`);
    }

    if (isIOS || isAndroid) {
      return this.native.displayNotification(options).then((): string => {
        return options.id as string;
      });
    }

    return Promise.resolve('');
  };

  public openAlarmPermissionSettings = (): Promise<void> => {
    if (isAndroid) {
      return this.native.openAlarmPermissionSettings();
    }

    return Promise.resolve();
  };

  public createTriggerNotification = (
    notification: Notification,
    trigger: Trigger,
  ): Promise<string> => {
    let options: Notification;
    let triggerOptions: Trigger;

    try {
      options = validateNotification(notification);
    } catch (e: any) {
      throw new Error(`notifee.createTriggerNotification(*) ${e.message}`);
    }

    try {
      triggerOptions = validateTrigger(trigger);
    } catch (e: any) {
      throw new Error(`notifee.createTriggerNotification(*) ${e.message}`);
    }

    if (isIOS || isAndroid) {
      return this.native.createTriggerNotification(options, triggerOptions).then((): string => {
        return options.id as string;
      });
    }

    return Promise.resolve('');
  };

  public getChannel = (channelId: string): Promise<NativeAndroidChannel | null> => {
    if (!isString(channelId)) {
      throw new Error("notifee.getChannel(*) 'channelId' expected a string value.");
    }

    if (isAndroid && this.native.ANDROID_API_LEVEL >= 26) {
      return this.native.getChannel(channelId);
    }

    return Promise.resolve(null);
  };

  public getChannels = (): Promise<NativeAndroidChannel[]> => {
    if (isAndroid && this.native.ANDROID_API_LEVEL >= 26) {
      return this.native.getChannels();
    }

    return Promise.resolve([]);
  };

  public getChannelGroup = (channelGroupId: string): Promise<NativeAndroidChannelGroup | null> => {
    if (!isString(channelGroupId)) {
      throw new Error("notifee.getChannelGroup(*) 'channelGroupId' expected a string value.");
    }

    if (isAndroid || this.native.ANDROID_API_LEVEL >= 26) {
      return this.native.getChannelGroup(channelGroupId);
    }

    return Promise.resolve(null);
  };

  public getChannelGroups = (): Promise<NativeAndroidChannelGroup[]> => {
    if (isAndroid || this.native.ANDROID_API_LEVEL >= 26) {
      return this.native.getChannelGroups();
    }

    return Promise.resolve([]);
  };

  public getInitialNotification = (): Promise<InitialNotification | null> => {
    if (isIOS || isAndroid) {
      return this.native.getInitialNotification();
    }

    return Promise.resolve(null);
  };

  public onBackgroundEvent = (observer: (event: Event) => Promise<void>): void => {
    if (!isFunction(observer)) {
      throw new Error("notifee.onBackgroundEvent(*) 'observer' expected a function.");
    }

    backgroundEventHandler = observer;
  };

  public onForegroundEvent = (observer: (event: Event) => void): (() => void) => {
    if (!isFunction(observer)) {
      throw new Error("notifee.onForegroundEvent(*) 'observer' expected a function.");
    }

    const subscriber = this.emitter.addListener(
      kReactNativeNotifeeNotificationEvent,
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore See https://github.com/facebook/react-native/pull/36462
      ({ type, detail }) => {
        observer({ type, detail });
      },
    );

    return (): void => {
      subscriber.remove();
    };
  };

  public openNotificationSettings = (channelId?: string): Promise<void> => {
    if (!isUndefined(channelId) && !isString(channelId)) {
      throw new Error("notifee.openNotificationSettings(*) 'channelId' expected a string value.");
    }

    if (isAndroid) {
      return this.native.openNotificationSettings(channelId || null);
    }

    return Promise.resolve();
  };

  public requestPermission = (
    permissions: IOSNotificationPermissions = {},
  ): Promise<NotificationSettings> => {
    if (isAndroid) {
      return this.native
        .requestPermission()
        .then(
          ({
            authorizationStatus,
            android,
          }: Pick<NotificationSettings, 'authorizationStatus' | 'android'>) => {
            return {
              authorizationStatus,
              android,
              ios: {
                alert: 1,
                badge: 1,
                criticalAlert: 1,
                showPreviews: 1,
                sound: 1,
                carPlay: 1,
                lockScreen: 1,
                announcement: 1,
                notificationCenter: 1,
                inAppNotificationSettings: 1,
                authorizationStatus,
              },
              web: {},
            };
          },
        );
    }

    if (isIOS) {
      let options: IOSNotificationPermissions;
      try {
        options = validateIOSPermissions(permissions);
      } catch (e: any) {
        throw new Error(`notifee.requestPermission(*) ${e.message}`);
      }

      return this.native
        .requestPermission(options)
        .then(
          ({
            authorizationStatus,
            ios,
          }: Pick<NotificationSettings, 'authorizationStatus' | 'ios'>) => {
            return {
              authorizationStatus,
              ios,
              android: {
                alarm: AndroidNotificationSetting.ENABLED,
              },
              web: {},
            };
          },
        );
    }

    // assume web
    return Promise.resolve({
      authorizationStatus: AuthorizationStatus.NOT_DETERMINED,
      android: {
        alarm: AndroidNotificationSetting.ENABLED,
      },
      ios: {
        alert: 1,
        badge: 1,
        criticalAlert: 1,
        showPreviews: 1,
        sound: 1,
        carPlay: 1,
        lockScreen: 1,
        announcement: 1,
        notificationCenter: 1,
        inAppNotificationSettings: 1,
        authorizationStatus: AuthorizationStatus.NOT_DETERMINED,
      },
      web: {},
    });
  };

  public registerForegroundService(runner: (notification: Notification) => Promise<void>): void {
    if (!isFunction(runner)) {
      throw new Error("notifee.registerForegroundService(_) 'runner' expected a function.");
    }

    if (isAndroid) {
      registeredForegroundServiceTask = runner;
    }

    return;
  }

  public setNotificationCategories = (categories: IOSNotificationCategory[]): Promise<void> => {
    if (!isIOS) {
      return Promise.resolve();
    }

    if (!isArray(categories)) {
      throw new Error(
        "notifee.setNotificationCategories(*) 'categories' expected an array of IOSCategory.",
      );
    }

    const options = [];
    try {
      for (let i = 0; i < categories.length; i++) {
        options[i] = validateIOSCategory(categories[i]);
      }
    } catch (e: any) {
      throw new Error(
        `notifee.setNotificationCategories(*) 'categories' a category is invalid: ${e.message}`,
      );
    }

    return this.native.setNotificationCategories(categories);
  };

  public getNotificationCategories = (): Promise<IOSNotificationCategory[]> => {
    if (!isIOS) {
      return Promise.resolve([]);
    }

    return this.native.getNotificationCategories();
  };

  public getNotificationSettings = (): Promise<NotificationSettings> => {
    if (isAndroid) {
      return this.native
        .getNotificationSettings()
        .then(
          ({
            authorizationStatus,
            android,
          }: Pick<NotificationSettings, 'authorizationStatus' | 'android'>) => {
            return {
              authorizationStatus,
              android,
              ios: {
                alert: 1,
                badge: 1,
                criticalAlert: 1,
                showPreviews: 1,
                sound: 1,
                carPlay: 1,
                lockScreen: 1,
                announcement: 1,
                notificationCenter: 1,
                inAppNotificationSettings: 1,
                authorizationStatus,
              },
              web: {},
            };
          },
        );
    }

    if (isIOS) {
      return this.native
        .getNotificationSettings()
        .then(
          ({
            authorizationStatus,
            ios,
          }: Pick<NotificationSettings, 'authorizationStatus' | 'ios'>) => {
            return {
              authorizationStatus,
              ios,
              android: {
                alarm: AndroidNotificationSetting.ENABLED,
              },
            };
          },
        );
    }

    // assume web
    return Promise.resolve({
      authorizationStatus: AuthorizationStatus.NOT_DETERMINED,
      android: {
        alarm: AndroidNotificationSetting.ENABLED,
      },
      ios: {
        alert: 1,
        badge: 1,
        criticalAlert: 1,
        showPreviews: 1,
        sound: 1,
        carPlay: 1,
        lockScreen: 1,
        announcement: 1,
        notificationCenter: 1,
        inAppNotificationSettings: 1,
        authorizationStatus: AuthorizationStatus.NOT_DETERMINED,
      },
      web: {},
    });
  };

  public getBadgeCount = (): Promise<number> => {
    if (!isIOS) {
      return Promise.resolve(0);
    }

    return this.native.getBadgeCount();
  };

  public setBadgeCount = (count: number): Promise<void> => {
    if (!isIOS) {
      return Promise.resolve();
    }

    if (!isNumber(count) || count < 0) {
      throw new Error("notifee.setBadgeCount(*) 'count' expected a number value greater than 0.");
    }

    return this.native.setBadgeCount(Math.round(count));
  };

  public incrementBadgeCount = (incrementBy?: number): Promise<void> => {
    if (!isIOS) {
      return Promise.resolve();
    }

    let value = 1;
    if (!isUndefined(incrementBy)) {
      if (!isNumber(incrementBy) || incrementBy < 1) {
        throw new Error(
          "notifee.decrementBadgeCount(*) 'incrementBy' expected a number value greater than 1.",
        );
      }

      value = incrementBy;
    }

    return this.native.incrementBadgeCount(Math.round(value));
  };

  public decrementBadgeCount = (decrementBy?: number): Promise<void> => {
    if (!isIOS) {
      return Promise.resolve();
    }

    let value = 1;
    if (!isUndefined(decrementBy)) {
      if (!isNumber(decrementBy) || decrementBy < 1) {
        throw new Error(
          "notifee.decrementBadgeCount(*) 'decrementBy' expected a number value greater than 1.",
        );
      }

      value = decrementBy;
    }

    return this.native.decrementBadgeCount(Math.round(value));
  };

  public isBatteryOptimizationEnabled = (): Promise<boolean> => {
    if (!isAndroid) {
      return Promise.resolve(false);
    }

    return this.native.isBatteryOptimizationEnabled();
  };

  public openBatteryOptimizationSettings = (): Promise<void> => {
    if (!isAndroid) {
      return Promise.resolve();
    }
    return this.native.openBatteryOptimizationSettings();
  };

  public getPowerManagerInfo = (): Promise<PowerManagerInfo> => {
    if (!isAndroid) {
      // only Android supports this, so instead we
      // return a dummy response to allow the power manager
      // flow work the same on all platforms
      return Promise.resolve({
        manufacturer: Platform.OS,
        activity: null,
      } as PowerManagerInfo);
    }

    return this.native.getPowerManagerInfo();
  };

  public openPowerManagerSettings = (): Promise<void> => {
    if (!isAndroid) {
      return Promise.resolve();
    }
    return this.native.openPowerManagerSettings();
  };

  public stopForegroundService = (): Promise<void> => {
    if (!isAndroid) {
      return Promise.resolve();
    }
    return this.native.stopForegroundService();
  };

  public hideNotificationDrawer = (): void => {
    if (!isAndroid) {
      return;
    }
    return this.native.hideNotificationDrawer();
  };
}
