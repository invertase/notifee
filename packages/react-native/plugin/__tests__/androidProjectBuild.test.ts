import fs from 'fs/promises';
import path from 'path';

import { setMavenRepository } from '../src/withNotifeeProjectGradlePlugin';

describe('[Android] project build.gradle test', () => {
  it('applies changes to project build.grade', async () => {
    let projectGradle = await fs.readFile(
      path.resolve(__dirname, '../fixtures/project.build.gradle'),
      {
        encoding: 'utf-8',
      },
    );

    projectGradle = setMavenRepository(projectGradle);

    expect(projectGradle).toMatchSnapshot();

    console.log(projectGradle);
  });
});
