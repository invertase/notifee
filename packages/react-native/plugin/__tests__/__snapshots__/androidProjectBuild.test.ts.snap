// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Android] project build.gradle test applies changes to project build.grade 1`] = `
"// Top-level build file where you can add configuration options common to all sub-projects/modules.

buildscript {
    ext {
        buildToolsVersion = \\"29.0.3\\"
        minSdkVersion = 21
        compileSdkVersion = 30
        targetSdkVersion = 30
    }
    repositories {
        google()
        jcenter()
    }
    dependencies {
        classpath(\\"com.android.tools.build:gradle:4.1.0\\")

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        mavenLocal()
        maven { url \\"$rootDir/../node_modules/@notifee/react-native/android/libs\\" }
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            url(\\"$rootDir/../node_modules/react-native/android\\")
        }
        maven {
            // Android JSC is installed from npm
            url(\\"$rootDir/../node_modules/jsc-android/dist\\")
        }

        google()
        jcenter()
        maven { url 'https://www.jitpack.io' }
    }
}
"
`;
